lib/ruby_smb/server/server_client/negotiation.rb

```diff
@@ -0,0 +1,153 @@
+require 'securerandom'
+
+module RubySMB
+  class Server
+    class ServerClient
+      module Negotiation
+        #
+        # Handle an SMB negotiation request. Once negotiation is complete, the state will be updated to :session_setup.
+        # At this point the @dialect will have been set along with other dialect-specific values.
+        #
+        # @param [String] raw_request the negotiation request to process
+        def handle_negotiate(raw_request)
+          response = nil
+          case raw_request[0...4].unpack1('L>')
+          when RubySMB::SMB1::SMB_PROTOCOL_ID
+            request = SMB1::Packet::NegotiateRequest.read(raw_request)
+            response = do_negotiate_smb1(request) if request.is_a?(SMB1::Packet::NegotiateRequest)
+          when RubySMB::SMB2::SMB2_PROTOCOL_ID
+            request = SMB2::Packet::NegotiateRequest.read(raw_request)
+            response = do_negotiate_smb2(request) if request.is_a?(SMB2::Packet::NegotiateRequest)
+          end
+
+          if response.nil?
+            disconnect!
+          else
+            send_packet(response)
+          end
+
+          nil
+        end
+
+        def do_negotiate_smb1(request)
+          if request.dialects.map(&:dialect_string).include?(Client::SMB1_DIALECT_SMB2_WILDCARD)
```

Comment:
I found an issue while testing dialect selection with SMBv1. I used this script to test:
```ruby
#!/usr/bin/ruby

require 'bundler/setup'
require 'ruby_smb'
require 'ruby_smb/gss/provider/ntlm'

ntlm_provider = RubySMB::Gss::Provider::NTLM.new(allow_anonymous: true, default_domain: '')
ntlm_provider.put_account('RubySMB', 'password')

server = RubySMB::Server.new(
  gss_provider: ntlm_provider
)
server.dialects = ['NT LM 0.12']

puts "server is running"
server.run do
  puts "received connection"
  true
end
```

The issue is that, most of the time, the client sends a `SMB1_DIALECT_SMB2_WILDCARD` first. So, if the server is setup to only accept SMBv1, like in the example above, it will never select this protocol version and always return a SMBv2 response with `SMB2::SMB2_WILDCARD_REVISION`. It is not possible to force the client to use SMBv1.

I think we should check if the server supports any SMBv2 dialect before sending a wild card SMBv2 response. Something along these lines:
```suggestion
          smb2_dialects = @server.dialects.select { |dialect| Dialect[dialect].order == Dialect::ORDER_SMB2 }
          if request.dialects.map(&:dialect_string).include?(Client::SMB1_DIALECT_SMB2_WILDCARD) && !smb2_dialects.empty?
```

Reply:
Great catch! You're right, I'll make that change. I don't think I tested only SMB1.